lib/streamlit/elements/vega_charts.py

```diff
@@ -185,12 +204,70 @@ def id_transform(data) -> dict[str, str]:
         with alt.data_transformers.enable("id"):  # type: ignore[attr-defined,unused-ignore]
             chart_dict = altair_chart.to_dict()
 
-    # Put datasets back into the chart dict but note how they weren't
-    # transformed.
+    # Put datasets back into the chart dict:
     chart_dict["datasets"] = datasets
     return chart_dict
 
 
+def _reset_counter_pattern(prefix: str, vega_spec: str) -> str:
+    """Altair uses a global counter for unnamed parameters and views.
+    We need to reset these counters on a spec-level to make the
+    spec stable across reruns and avoid changes to the element ID.
+    """
+    pattern = re.compile(rf'"{prefix}\d+"')
+    # Get all matches without duplicates in order of appearance.
+    # Using a set here would not guarantee the order of appearance,
+    # which might lead to different replacements on each run.
+    if matches := list(dict.fromkeys(pattern.findall(vega_spec))):
+        # Add a prefix to the replacement to avoid
+        # replacing instances that already have been replaced before.
+        replacement_prefix = "__replace_prefix_o9hd101n22e1__"
+
+        # Replace all matches with a counter starting from 1
+        # We start from 1 to imitate the altair behavior.
+        for counter, match in enumerate(matches, start=1):
+            vega_spec = vega_spec.replace(
+                match, f'"{replacement_prefix}{prefix}{counter}"'
+            )
+
+        # Remove the prefix again from all replacements:
+        vega_spec = vega_spec.replace(replacement_prefix, "")
+    return vega_spec
+
+
+def _stabilize_vega_json_spec(vega_spec: str) -> str:
+    """Makes the chart spec stay stable across reruns.
+
+    Altair auto creates names for unnamed parameters & views. It uses a global counter
+    for the naming which will result in a different spec on every rerun.
+    In Streamlit, we need the spec to be stable across reruns to prevent the chart
+    from getting a new identity. So we need to replace the names with counter with a stable name.
+
+    Parameter counter:
+    https://github.com/vega/altair/blob/f345cd9368ae2bbc98628e9245c93fa9fb582621/altair/vegalite/v5/api.py#L196
+
+    View counter:
+    https://github.com/vega/altair/blob/f345cd9368ae2bbc98628e9245c93fa9fb582621/altair/vegalite/v5/api.py#L2885
+
+    This is temporary solution waiting for a fix for this issue:
+    https://github.com/vega/altair/issues/3416
+    """
```

Comment:
nit: it might be a good idea to quickly summarize the other solutions we talked about offline and why this one was chosen, e.g.

```
Other solutions we considered:
- working on the JSON object: this would require to iterate through the object and do the same kind of replacement; though we would need to know the structure and since we need the spec in String-format anyways, we deemed that executing the replacement on the String is the better alternative
- resetting the counter: the counter is incremented already when the chart object is created (see this GitHub issue comment XYZ), so it would be too late here to reset the counter with a thread-lock to prevent interference between sessions
```

Reply:
Also it might make sense to add that you think it's not a good idea to make a `key` field mandatory on `st.chart` when wanting to have interactivity as this would behave differently to other widgets.

And perhaps one sentence about FowardMsgCache not working efficiently when the spec changes on every rerun.

Reply:
I added it to the docstring comment